channeldb/kvdb/etcd/commit_queue_test.go

```diff
@@ -68,48 +69,51 @@ func TestCommitQueue(t *testing.T) {
 	defer cancel()
 
 	wg.Add(numCommits)
-	t1 := time.Now()
+
+	var block sync.Mutex
 
 	// Tx1: reads: key1, key2, writes: key3, conflict: none
-	q.Add(
-		commit("free", true),
+	// Since we simulate that the txn takes a long time, we'll add in a
+	// new goroutine and wait for the txn commit to start execution.
+	commitTx1 := make(chan struct{})
+	// Block the commit by holding the mutex.
+	block.Lock()
+	go q.Add(
+		commit("free", commitTx1, &block),
 		makeReadSet([]string{"key1", "key2"}),
 		makeWriteSet([]string{"key3"}),
 	)
-	// Tx2: reads: key1, key2, writes: key3, conflict: Tx1
-	q.Add(
-		commit("blocked1", false),
-		makeReadSet([]string{"key1", "key2"}),
-		makeWriteSet([]string{"key3"}),
-	)
-	// Tx3: reads: key1, writes: key4, conflict: none
+	<-commitTx1
+
+	// Tx2: reads: key1, key5, writes: key3, conflict: Tx1 (key3)
+	// We don't expect queue add to block as this txn will queue up after
+	// tx1.
 	q.Add(
-		commit("free", true),
-		makeReadSet([]string{"key1", "key2"}),
-		makeWriteSet([]string{"key4"}),
+		commit("blocked1", nil, nil),
+		makeReadSet([]string{"key1"}),
+		makeWriteSet([]string{"key3", "key5"}),
 	)
-	// Tx4: reads: key2, writes: key4 conflict: Tx3
+
+	// Tx3: reads: key2, key5, writes: key4 conflict: Tx2 (key5)
+	// We don't expect queue add to block as this txn will queue up after
+	// tx2.
 	q.Add(
-		commit("blocked2", false),
-		makeReadSet([]string{"key2"}),
+		commit("blocked2", nil, nil),
+		makeReadSet([]string{"key2", "key5"}),
 		makeWriteSet([]string{"key4"}),
 	)
 
+	// Allow Tx1 to continue with the commit.
+	block.Unlock()
```

Comment:
nit: might be simpler to use a `ready` channel, e.g.:
```golang
commit := func(...) {
        ...
        if ready != nil {
              <-ready
        }
}
...
ready := make(chan struct{})
...
commit(_, _, ready)
...
close(ready)
```

This approach will also work with more than one "free" transaction

Reply:
See my other comment. We can use a channel, but in the test example it's not possible to submit overlapping conflicting/non-conflicting txns without introducing a deadlock in the test.

Reply:
I agree that using a channel to signal that the commit can continue is better than a mutex